### DISCUSSION TITLE
update ref.read(provider) in repository class

### DISCUSSION THREAD

**Discussion by vchrisb** [rrousselGit/riverpod#3286] [Q&A] [answered]
Hi,

I think my code is similar to the example in the docs [Can I read a provider without listening to it?](https://riverpod.dev/docs/concepts/combining_providers#can-i-read-a-provider-without-listening-to-it)

But I do not get the expected outcome.

I do have a `LocationRepository` where I pass the `ref` in to be able to read another provider `mapVisibleBoundsProvider`.
I'm using `ref.read`, as I don't want my `locationRepositoryProvider` to issue the netwok call again, whenever  `mapVisibleBoundsProvider` (AutoDisposeNotifierProvider) is updated.
I only want my `locationsProvider` to be refreshed when calling ref.refresh(locationsProvider.future) in a widget.

My problem is, that the `bounds = ref.read(mapVisibleBoundsProvider);` is not being reevaluated when calling `ref.refresh(locationsProvider.future) ` . The `mapVisibleBoundsProvider` is successful updated from some widget but the value of `bounds` never changes.

What am I doing wrong?

```
// "location" repository
@riverpod
LocationRepository locationRepository(LocationRepositoryRef ref) =>
    LocationRepository(ref: ref);

// "location" provider
@riverpod
Future<List<LocationSlim>> locations(LocationsRef ref) {
  final cancelToken = CancelToken();
  return ref
      .watch(locationRepositoryProvider)
      .locations(cancelToken: cancelToken);
}

// repository
class LocationRepository {
  LocationRepository({required this.ref});
  final Ref ref;

  Future<List<LocationSlim>> locations({CancelToken? cancelToken}) async {

    final bounds = ref.read(mapVisibleBoundsProvider);
    
    ... use "bounds" in a network call ...```

**Comment by AhmedLSayed9** [comment]
How do you expect `final bounds = ref.read(mapVisibleBoundsProvider);` to re-evaluate exactly? and maybe showing its code can help.

**Reply by vchrisb** [reply]
Thank you @AhmedLSayed9 for your reply!

I expect the `final bounds = ref.read(mapVisibleBoundsProvider);` to be  re-evaluate whenever I do call `ref.refresh(locationsProvider.future)`.

I'll use the value of `bounds` to construct some queryparameters. that'll use in the network request.
```
final bbox = {
      'in_bbox': "${bounds.west},${bounds.south},${bounds.east},${bounds.north}"
    };
```

**Reply by AhmedLSayed9** [reply]
Okay, You're manually updating `mapVisibleBoundsProvider` from some widget, but are you listening to `mapVisibleBoundsProvider` somewhere?

**Reply by vchrisb** [reply]
Yes, I'm updating `mapVisibleBoundsProvider` from some widget 
Something like:
```
// across some widgets
onMove: () => ref.read(mapVisibleBoundsProvider.notifier).update(<<new bounds>>),
onTap: () => ref.refresh(locationsProvider.future),
```

**Reply by vchrisb** [reply]
So whenever I move my map, I do update the bounds, stored in `mapVisibleBoundsProvider`, but only when tapping, I do call ` ref.refresh(locationsProvider.future)`, expecting the Location repository class to get the latest value of `mapVisibleBoundsProvider`.

**Reply by AhmedLSayed9** [reply]
ref.read doesn't add a listener for your provider.

so, when you do:
```
ref.read(mapVisibleBoundsProvider.notifier).update(<<new bounds>>)
```
Your provider is being built and updated with new bounds and then immediately disposed. When you read it again, you read the initial value.

You need to listen for the provider somewhere using ref.watch or ref.listen.

i.e If this is used at the map screen lifecycle and you don't care about updates from `mapVisibleBoundsProvider`, you can add the following at the start of your map screen:
```dart
listen(mapVisibleBoundsProvider, (_, __) {});
```

**Reply by vchrisb** [reply, marked as answer]
Thank you @AhmedLSayed9 🥇 
You send me in the right direction! 

I've annotated my `MapVisibleBounds` Class with `@Riverpod(keepAlive: true)`.
Now I have a `NotifierProvider` instead of an `AutoDisposeNotifierProvider`!

```
@Riverpod(keepAlive: true)
class MapVisibleBounds extends _$MapVisibleBounds {
  @override
  LatLngBounds build() {
    return LatLngBounds(const LatLng(0, 0), const LatLng(0, 0));
  }

  void update(LatLngBounds visibleBounds) {
    state = visibleBounds;
  }
}
````

**Reply by AhmedLSayed9** [reply]
I'd prefer using auto-dispose notifier and listening to the provider at the correct widget instead.
By using `keepAlive`, when you leave the map screen and enter again, the updated `LatLngBounds` will be used and it will not reset.

**Reply by vchrisb** [reply]
I'll keep that in mind.
